### PR TITLE
crimson/osd: minor bug fixes misc

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/crimson/osd/ops_executer.h [hunk 1]
```diff
@@ -504,6 +504,7 @@ OpsExecuter::flush_changes_n_do_ops_effects(
     ceph_assert(want_mutate);
   }
 
+  apply_stats();
```
Comment (Matan-B) [inline src/crimson/osd/ops_executer.h:507]: Can you please explain why moving OpsExecutor::obc is relevant to apply_stats?

Comment (xxhdx1985126) [reply src/crimson/osd/ops_executer.h:507]: `apply_stats` also tries to access `OpsExecutor::obc`, so it has to called before we "move" `OpsExecutor::obc`. I'll add comments to the code.

Comment (Matan-B) [reply src/crimson/osd/ops_executer.h:507]: Not sure where it tries to access it, I might be missing something here. To me it looks we don't actually use obc.

Comment (xxhdx1985126) [reply src/crimson/osd/ops_executer.h:507]: Yeah, I think I was wrong about this, will re-examine the case.

Comment (xxhdx1985126) [reply src/crimson/osd/ops_executer.h]: @Matan-B Hi, matan, it turned out that the case was caused by other conditions and only exist in our test version. I've just removed that commit, please take a look again, thanks:-)

**Review by Matan-B — APPROVED** [review]

File: src/crimson/osd/ops_executer.h (hunk 1 above)
Comment (xxhdx1985126) [reply src/crimson/osd/ops_executer.h:507]: @Matan-B Sorry, I was wrong again, apply_stats does access obc, please see the following backtrace:
```
crimson-osd: /home/xuxuehan/src/ceph/clang-build/boost/include/boost/smart_ptr/intrusive_ptr.hpp:201: T *boost::intrusive_ptr<crimson::osd::ObjectContext>::operator->() const [T = crimson::osd::ObjectContext]: Assertion `px != 0' failed.
Aborting on shard 0.
Backtrace:
Reactor stalled for 3359 ms on shard 0, in scheduling group main. Backtrace: 0x78a43d7 0x78a422c 0x783c1b6 0x783acd0 0x783abb0 0x783ae4b 0x783c586 0x4251f 0x174627 0x5b012b0 0x5b017fd 0x5b028b9 0x5b011a6 0x5b00f91 0x5b00be2 0x5b00a4b 0x5b00978 0x5b000c4 0x5affd82 0x5affcc5 0x4251f 0x969fb 0x42475 0x287f2 0x2871a 0x39e95 0x4dc4091 0x4e3bc7b 0x4dc87a4 0x4fdebfc 0x4dba57e 0x4e1cff7 0x4e1cd94 0x7845d61 0x784731a 0x7848352 0x7847734 0x77aa304 0x77a9746 0x4b74c4e 0x29d8f 0x29e3f 0x4ad2fc4
Reactor stalled for 6383 ms on shard 0, in scheduling group main. Backtrace: 0x78a43d7 0x78a422c 0x783c1b6 0x783acd0 0x783abb0 0x783ae4b 0x783c586 0x4251f 0x1745e5 0x5b012b0 0x5b017fd 0x5b01503 0x5b013f9 0x5b013c0 0x5b01131 0x5b00f21 0x5b00be2 0x5b00a4b 0x5b00978 0x5b000c4 0x5affd82 0x5affcc5 0x4251f 0x969fb 0x42475 0x287f2 0x2871a 0x39e95 0x4dc4091 0x4e3bc7b 0x4dc87a4 0x4fdebfc 0x4dba57e 0x4e1cff7 0x4e1cd94 0x7845d61 0x784731a 0x7848352 0x7847734 0x77aa304 0x77a9746 0x4b74c4e 0x29d8f 0x29e3f 0x4ad2fc4
 0# pthread_kill in /lib/x86_64-linux-gnu/libc.so.6
 1# raise in /lib/x86_64-linux-gnu/libc.so.6
 2# abort in /lib/x86_64-linux-gnu/libc.so.6
 3# 0x00007FD82A38D71B in /lib/x86_64-linux-gnu/libc.so.6
 4# 0x00007FD82A39EE96 in /lib/x86_64-linux-gnu/libc.so.6
 5# boost::intrusive_ptr<crimson::osd::ObjectContext>::operator->() const in /home/xuxuehan/src/ceph/clang-build/bin/crimson-osd
 6# crimson::osd::OpsExecuter::get_object_info() const in /home/xuxuehan/src/ceph/clang-build/bin/crimson-osd
 7# crimson::osd::OpsExecuter::get_target() const in /home/xuxuehan/src/ceph/clang-build/bin/crimson-osd
 8# crimson::osd::OpsExecuter::apply_stats() in /home/xuxuehan/src/ceph/clang-build/bin/crimson-osd
 9# crimson::interruptible::interruptible_future_detail<crimson::osd::IOInterruptCondition, seastar::future<std::tuple<crimson::interruptible::interruptible_future_detail<crimson::osd::IOInterruptCondition, seastar::future<void> >, crimson::interruptible::interruptible_future_detail<crimson::osd::IOInterruptCondition, seastar::future<void> > > > > crimson::osd::OpsExecuter::flush_changes_n_do_ops_effects<crimson::osd::PG::submit_executer(crimson::osd::OpsExecuter&&, std::vector<OSDOp, std::allocator<OSDOp> > const&)::$_0>(std::vector<OSDOp, std::allocator<OSDOp> > const&, SnapMapper&, OSDriver&, crimson::osd::PG::submit_executer(crimson::osd::OpsExecuter&&, std::vector<OSDOp, std::allocator<OSDOp> > const&)::$_0) && [clone .resume] at pg.cc:?
10# std::__n4861::coroutine_handle<crimson::internal::promise_base<crimson::interruptible::interruptor<crimson::osd::IOInterruptCondition>, void, std::tuple<crimson::interruptible::interruptible_future_detail<crimson::osd::IOInterruptCondition, seastar::future<void> >, crimson::interruptible::interruptible_future_detail<crimson::osd::IOInterruptCondition, seastar::future<void> > > > >::resume() const in /home/xuxuehan/src/ceph/clang-build/bin/crimson-osd
11# crimson::internal::promise_base<crimson::interruptible::interruptor<crimson::osd::IOInterruptCondition>, void, std::tuple<crimson::interruptible::interruptible_future_detail<crimson::osd::IOInterruptCondition, seastar::future<void> >, crimson::interruptible::interruptible_future_detail<crimson::osd::IOInterruptCondition, seastar::future<void> > > >::run_and_dispose() in /home/xuxuehan/src/ceph/clang-build/bin/crimson-osd
12# seastar::reactor::run_tasks(seastar::reactor::task_queue&) at :?
```

I've just added the commit back, please take a look, again. Thanks.

Comment (Matan-B) [reply src/crimson/osd/ops_executer.h:507]: c5a05faed82129daeb283cfdfda3e4d7e42c1e87 moved the apply_stats call there in order to allow for read ops to share stats as well.
By moving the call up we'll apply the stats before `flush_clone_metadata`, I think this should be ok as we adjust the clone's delta_stats in OpsExecuter ctor.

**Comment by Matan-B** [comment]
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-nov-17

**Pull request merged**